src/main/java/org/commonjava/util/sidecar/util/ProxyStreamingOutput.java

```diff
@@ -16,52 +16,135 @@
+import okhttp3.ResponseBody;
+import okio.BufferedSource;
 import org.apache.commons.io.output.CountingOutputStream;
+import org.commonjava.util.sidecar.model.TrackedContentEntry;
+import org.commonjava.util.sidecar.services.ReportService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.ws.rs.core.StreamingOutput;
+import javax.xml.bind.DatatypeConverter;
 import java.io.IOException;
-import java.io.InputStream;
 import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
 
 public class ProxyStreamingOutput
                 implements StreamingOutput
 {
     private final Logger logger = LoggerFactory.getLogger( getClass() );
 
-    private final InputStream bodyStream;
+    private static final String MD5 = "MD5";
+
+    private static final String SHA1 = "SHA-1";
+
+    private static final String SHA256 = "SHA-256";
+
+    private static final String[] DIGESTS = { MD5, SHA1, SHA256 };
+
+    private static final long bufSize = 10 * 1024 * 1024;
+
+    private final ResponseBody responseBody;
+
+    private final TrackedContentEntry entry;
+
+    private final String serviceOrigin;
+
+    private final String indyOrigin;
+
+    private final ReportService reportService;
 
     private final OtelAdapter otel;
 
-    public ProxyStreamingOutput( InputStream bodyStream, OtelAdapter otel )
+    private final Map<String, MessageDigest> digests = new HashMap<>();
+
+    public ProxyStreamingOutput( ResponseBody responseBody, TrackedContentEntry entry, String serviceOrigin,
+                                 String indyOrigin, ReportService reportService, OtelAdapter otel )
     {
-        this.bodyStream = bodyStream;
+        this.responseBody = responseBody;
+        this.entry = entry;
+        this.serviceOrigin = serviceOrigin;
+        this.indyOrigin = indyOrigin;
+        this.reportService = reportService;
         this.otel = otel;
+
+        for ( String key : DIGESTS )
+        {
+            try
+            {
+                digests.put( key, MessageDigest.getInstance( key ) );
+            }
+            catch ( NoSuchAlgorithmException e )
+            {
+                logger.warn( "Bytes hash calculation failed for request. Cannot get digest of type: {}", key );
+            }
+        }
     }
 
     @Override
     public void write( OutputStream output ) throws IOException
     {
-        if ( bodyStream != null )
+        if ( responseBody != null )
         {
-            try
+            try (CountingOutputStream cout = new CountingOutputStream( output ))
             {
-                OutputStream out = output;
-                CountingOutputStream cout = new CountingOutputStream( out );
-                out = cout;
-                logger.trace( "Copying from: {} to: {}", bodyStream, out );
-                IOUtils.copy( bodyStream, out );
-
+                OutputStream out = cout;
```

Comment:
Is this out needed? Why not use cout directly here?

Reply:
@ligangty This is the same with the previously removed code, I think what John considered here might be out is used for writing, cout is used for bytes count.